smartsim/_core/launcher/dragon/dragonConnector.py

```diff
@@ -0,0 +1,447 @@
+                "experiment.\nIf the DragonConnector was started manually, "
+                "then the environment variable SMARTSIM_DRAGON_SERVER_PATH "
+                "should be set to an existing directory."
+            )
+
+    @property
+    def is_connected(self) -> bool:
+        return self._dragon_head_socket is not None
+
+    def _handshake(self, address: str) -> None:
+        self._dragon_head_socket = dragonSockets.get_secure_socket(
+            self._context, zmq.REQ, False
+        )
+        self._dragon_head_socket.connect(address)
+        try:
+            dragon_handshake = _assert_schema_type(
+                self.send_request(DragonHandshakeRequest()), DragonHandshakeResponse
+            )
+            self._dragon_head_pid = dragon_handshake.dragon_pid
+            logger.debug(
+                f"Successful handshake with Dragon server at address {address}"
+            )
+        except (zmq.ZMQError, zmq.Again) as e:
+            logger.debug(e)
+            self._dragon_head_socket.close()
+            self._dragon_head_socket = None
+
+            raise SmartSimError(
+                f"Unsuccessful handshake with Dragon server at address {address}"
+            ) from e
+
+    def _set_timeout(self, timeout: int) -> None:
+        self._context.setsockopt(zmq.SNDTIMEO, value=timeout)
+        self._context.setsockopt(zmq.RCVTIMEO, value=timeout)
+        if self._authenticator is not None and self._authenticator.thread is not None:
+            try:
+                self._authenticator.thread.authenticator.zap_socket.setsockopt(
+                    zmq.SNDTIMEO, timeout
+                )
+                self._authenticator.thread.authenticator.zap_socket.setsockopt(
+                    zmq.RCVTIMEO, timeout
+                )
+            except zmq.ZMQError:
+                pass
+
+    def ensure_connected(self) -> None:
+        if not self.is_connected:
+            self.connect_to_dragon()
+        if not self.is_connected:
+            raise SmartSimError("Could not connect to Dragon server")
+
+    def _get_new_authenticator(self) -> None:
+        if self._authenticator is not None:
+            if self._authenticator.thread is not None:
+                try:
+                    logger.debug("Closing ZAP socket")
+                    self._authenticator.thread.authenticator.zap_socket.close()
+                except Exception as e:
+                    logger.debug(f"Could not close ZAP socket, {e}")
+            try:
+                self._authenticator.stop()
+            except zmq.Again:
+                logger.debug("Could not stop authenticator")
+        try:
+            self._authenticator = dragonSockets.get_authenticator(self._context)
+            return
+        except RuntimeError as e:
+            logger.error("Could not get authenticator")
+            raise e from None
+
+    # pylint: disable-next=too-many-statements,too-many-locals
+    def connect_to_dragon(self) -> None:
```

Comment:
Can we break down this big method now before it grows and presents `too-many-locals` linter complaints? A couple natural methods to extract might be L177-200 (e.g. `def connect_to_existing`) and L204-231 (e.g. `def build_cmd`).

Reply:
Done